.github/workflows/CICD.yaml

```diff
@@ -189,6 +191,58 @@ jobs:
           retention-days: 1
           if-no-files-found: error
 
+  Bandit:
+    name: SAST scanning
+    runs-on: ubuntu-latest
+    needs:
+      - Checkout
+
+    env:
+      PYTHON_VERSION: "3.10"
+
+    if: ${{( success() && !cancelled() ) }}
+
+    # FIXME Ignore errors while testing Bandit
+    continue-on-error: true
+
+    steps:
+      - name: Set up Python ${{ env.PYTHON_VERSION }}
+        uses: actions/setup-python@v5
+        id: install_python
+        with:
+          python-version: ${{ env.PYTHON_VERSION }}
+
+      - uses: actions/cache/restore@v4
+        name: Restore run cache
+        id: restore-run-cache
+        with:
+          key: ${{ needs.Checkout.outputs.cache-key-run }}
+          path: ${{ github.workspace }}
+          fail-on-cache-miss: true
+
+      - name: Run bandit scan and generate reports
+        run: |
+          echo Running bandit
+
+          VENV=.github-venv \
+          DEFAULT_TARGET=cicd \
+            make test-bandit
+
+      - name: Output bandit report artifact
+        uses: actions/upload-artifact@v4
+        if: ${{ always() }}
+        with:
+          name: bandit-sast-report
+          path: |
+            bandit.sarif
+          retention-days: 1
+          if-no-files-found: error
```

Comment:
Thanks for working through this - makes sense to me. I'm wondering if you think there could be a more explicit way to handle this? Of course, It's always difficult to write declaratively in GH actions.

Reply:
I've been pondering how to describe how I read and write these files, with regards to the flow of jobs and steps. I think it can be summarized as: each step is a response to the prior step in the whole of a state machine. Once one state is done, it can be ignored for consideration in future steps.

Therefore, I think the way toward more explicit steps is to hide as much complexity as possible. This allows us to think through fewer possibilities. In this case, much of the complexity of what is considered failure is hidden in the Makefile and does not need to be considered. This step can be read as "always upload the artifact regardless of what has already happened, but fail if the file doesn't exist for some reason, regardless of what that reason is." The following step, then, has no `if` criterion, and so the job if uploading failed.

Another possibility to make things more explicit is to _always_ include a criteria for steps, even if it just handles the implicit `success()` case.

---
With regards to considering what the step actually does, the Makefile clarifies a bit - there is a comment that says not to error if `bandit` errors. However, it does not specifically call out that `bandit` errors if any SAST problems are reported. What this means for the CI/CD pipeline is that, while testing, no SAST problems should halt CI/CD _however_ if `bandit` didn't create the report file for some reason _there is some other problem_ and the pipeline should stop. I suppose a way to make this more explicit would be to check for the file in `Run bandit scan and generate reports` and error if it's not there, but then we increase complexity and replicate this step's `if-no-files-found` condition.

Perhaps just adding a comment to the step would be sufficient to explain what's going on.

---
One other thing worth considering - as I have set up this pipeline, I'm familiar with my expectations and intent for the pipeline overall. Perhaps documentation on the pipeline would make these behaviors self-evident.